stake-pool/program/src/processor.rs

```diff
@@ -1410,37 +1435,134 @@ impl Processor {
-            reserve_stake_account_info.clone(),
-            withdraw_authority_info.clone(),
-            AUTHORITY_WITHDRAW,
-            stake_pool.stake_withdraw_bump_seed,
-            total_lamports,
-            transient_stake_account_info.clone(),
-        )?;
+                // split into ephemeral stake account
+                Self::stake_split(
+                    stake_pool_info.key,
+                    reserve_stake_account_info.clone(),
+                    withdraw_authority_info.clone(),
+                    AUTHORITY_WITHDRAW,
+                    stake_pool.stake_withdraw_bump_seed,
+                    total_lamports,
+                    ephemeral_stake_account_info.clone(),
+                )?;
 
-        // activate transient stake to validator
-        Self::stake_delegate(
-            transient_stake_account_info.clone(),
-            validator_vote_account_info.clone(),
-            clock_info.clone(),
-            stake_history_info.clone(),
-            stake_config_info.clone(),
-            withdraw_authority_info.clone(),
-            stake_pool_info.key,
-            AUTHORITY_WITHDRAW,
-            stake_pool.stake_withdraw_bump_seed,
-        )?;
+                // activate stake to validator
+                Self::stake_delegate(
+                    ephemeral_stake_account_info.clone(),
+                    validator_vote_account_info.clone(),
+                    clock_info.clone(),
+                    stake_history_info.clone(),
+                    stake_config_info.clone(),
+                    withdraw_authority_info.clone(),
+                    stake_pool_info.key,
+                    AUTHORITY_WITHDRAW,
+                    stake_pool.stake_withdraw_bump_seed,
+                )?;
+                ephemeral_stake_account_info
+            } else {
+                // if no ephemeral account is provided, split everything from the
+                // reserve account, into the transient stake account
+                reserve_stake_account_info
+            };
+
+        if validator_stake_info.transient_stake_lamports > 0 {
+            // transient stake exists, try to merge from the source account
+            Self::stake_merge(
+                stake_pool_info.key,
+                source_stake_account_info.clone(),
+                withdraw_authority_info.clone(),
+                AUTHORITY_WITHDRAW,
+                stake_pool.stake_withdraw_bump_seed,
+                transient_stake_account_info.clone(),
+                clock_info.clone(),
+                stake_history_info.clone(),
+                stake_program_info.clone(),
+            )?;
+        } else {
+            // no transient stake, split
+            let transient_stake_bump_seed = check_transient_stake_address(
+                program_id,
+                stake_pool_info.key,
+                transient_stake_account_info.key,
+                vote_account_address,
+                transient_stake_seed,
+            )?;
```

Comment:
we should move the call to `check_transient_stake_address` outside the if/else. as it stands, if `validator_stake_info.transient_stake_lamports > 0` is true, `transient_stake_account_info` is never validated. the pool can be griefed by creating an unrelated stake account with the same lockup/authority to merge into instead, resulting in a miscalculation in `validator_stake_info.transient_stake_lamports` at the end of this function, which could have unknown downstream effects

also for the record today i learned that merging stake accounts requires them to have the same authority, the missing check baited me into partway writing a poc for loss of funds before i learned this fact :sweat_smile:

Reply:
Great catch, thanks! I don't think you could steal funds, but you could definitely mess up the pool management if you spoof this.